core/src/inputs/advice/merkle_path_set.rs

```diff
@@ -0,0 +1,320 @@
+use super::{hasher, AdviceSetError, Felt, FieldElement, Word};
+use std::collections::BTreeMap;
+
+// MERKLE PATH SET
+// ================================================================================================
+
+/// A set of Merkle paths.
+///
+/// This struct is intended to be used as one of the variants of the MerkleSet enum.
+#[derive(Clone, Debug)]
+pub struct MerklePathSet {
+    root: Word,
+    total_depth: u32,
+    paths: BTreeMap<u64, Vec<Word>>,
+}
+
+impl MerklePathSet {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns an empty MerklePathSet.
+    pub fn new(depth: u32) -> Result<Self, AdviceSetError> {
+        let root = [Felt::ZERO; 4];
+        let paths = BTreeMap::<u64, Vec<Word>>::new();
+
+        Ok(Self {
+            root,
+            total_depth: depth,
+            paths,
+        })
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Adds the Merkle Path to the paths map.
+    /// index -- index of the node, which path we want to add
+    /// hash -- hash of the node itself
+    /// path -- Merkle Path of the node
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * Attempt to add a path to leaf that differs from the existing one.
+    pub fn add_path(
+        &mut self,
+        index: u64,
+        value: Word,
+        path: Vec<Word>,
+    ) -> Result<(), AdviceSetError> {
+        let depth = (path.len() + 1) as u32;
+        if depth != self.total_depth {
+            return Err(AdviceSetError::InvalidPathLength(self.total_depth, depth));
+        }
+
+        // actual number of node in tree
+        let pos = 2u64.pow(self.total_depth) + index;
+
+        // index of the leaf path in map. We store paths of neighboring leaves in one key-value pair
+        let half_pos = (pos / 2) as u64;
+
+        // Now we returning an error, if the new path is not the same as the one in the map
+        let mut extended_path = path;
+        if is_even(pos) {
+            extended_path.insert(0, value);
+        } else {
+            extended_path.insert(1, value);
+        }
+
+        if self.paths.get(&half_pos) == None {
+            let root_of_current_path = compute_root(&extended_path, depth, index);
+            if self.root == [Felt::ZERO; 4] {
+                self.root = root_of_current_path;
+            } else if self.root != root_of_current_path {
+                return Err(AdviceSetError::InvalidPath(extended_path));
+            }
+            self.paths.insert(half_pos, extended_path);
+        } else if *self.paths.get(&half_pos).unwrap() != extended_path {
+            return Err(AdviceSetError::InvalidPath(extended_path));
+        }
```

Comment:
You should probably be able to simplify this as:
```Rust
let root_of_current_path = compute_root(&extended_path, depth, index);
if self.root == [Felt::ZERO; 4] {
    self.root = root_of_current_path;
} else if self.root != root_of_current_path {
    return Err(AdviceSetError::InvalidPath(extended_path));
}
self.paths.insert(half_pos, extended_path);
```
Basically, if the path resolved to a different root, it would not be equal to the path stored for the same index.